jerry-core/include/jerryscript-port.h

```diff
@@ -16,233 +16,294 @@
- *                    filesystem.
- * @param out_size_p The opened file's size in bytes.
+ * @param file_name_p: Path that points to the source file in the filesystem.
+ * @param out_size_p: The opened file's size in bytes.
  *
- * @return the pointer to the buffer which contains the content of the file.
+ * @return pointer to the buffer which contains the content of the file.
  */
-uint8_t *jerry_port_read_source (const char *file_name_p, size_t *out_size_p);
+jerry_char_t *jerry_port_source_read (const char *file_name_p, jerry_size_t *out_size_p);
 
 /**
- * Frees the allocated buffer after the contents of the file are not needed
- * anymore.
+ * Free a source file buffer.
  *
- * @param buffer_p The pointer the allocated buffer.
+ * @param buffer_p: buffer allocated by jerry_port_source_read
+ */
+void jerry_port_source_free (jerry_char_t *buffer_p);
+
+/**
+ * jerry-port-fs @}
+ */
+
+/**
+ * @defgroup jerry-port-date Date API
+ * @{
  */
-void jerry_port_release_source (uint8_t *buffer_p);
 
 /**
- * Default module resolver.
+ * Get local time zone adjustment in milliseconds for the given input time.
+ *
+ * The argument is a time value representing milliseconds since unix epoch.
+ *
+ * Ideally, this function should satisfy the stipulations applied to LocalTZA
+ * in section 21.4.1.7 of the ECMAScript version 12.0, as if called with isUTC true.
  *
- * Note:
- *      This port function is only used when JERRY_MODULE_SYSTEM is enabled.
+ * This port function can be called by jerry-core when JERRY_BUILTIN_DATE is enabled.
+ * Otherwise this function is not used.
  *
- * @param specifier Module specifier string.
- * @param referrer Parent module.
- * @param user_p An unused pointer.
+ * @param unix_ms: time value in milliseconds since unix epoch
  *
- * @return A module object if resolving is successful, an error otherwise.
+ * @return local time offset in milliseconds applied to UTC for the given time value
  */
-jerry_value_t jerry_port_module_resolve (const jerry_value_t specifier, const jerry_value_t referrer, void *user_p);
+int32_t jerry_port_local_tza (double unix_ms);
 
 /**
- * Release known modules.
+ * Get the current system time in UTC.
  *
- * Note:
- *      This port function should be called by the user application when
- *      the module database is no longer needed.
+ * This port function is called by jerry-core when JERRY_BUILTIN_DATE is enabled.
+ * It can also be used in the implementing application to initialize the random number generator.
  *
- * @param realm If this argument is object, release only those modules,
- *              which realm value is equal to this argument.
+ * @return milliseconds since Unix epoch
  */
-void jerry_port_module_release (const jerry_value_t realm);
+double jerry_port_current_time (void);
 
 /**
- * @}
+ * jerry-port-date @}
+ */
+
+/**
+ * jerry-port @}
  */
 
 JERRY_C_API_END
 
 #endif /* !JERRYSCRIPT_PORT_H */
+
+/* vim: set fdm=marker fmr=@{,@}: */
```

Comment:
Does this are necessary?

Reply:
It's not necessary, however it does not hurt anyone, and may be helpful for some.

Reply:
Can this be in config file?
as `jerryscript-core.h` also need this.

Reply:
I almost wrote that we don't have anything like that in the current code base. Unfortunately, it is not completely true, as jerryscript-core.h already contains such a line. It has crept in somehow. IMHO it should not have.

Reply:
I don't really see what the fuss is about. 

It is a file specific comment for a specific editor. We have other options set here and there for other IDE-s as well. Bunch of git-ignores for a lot of editors, generated compile_commands.json to enable code completion in vscode, etc.

It is a simple comment, it has no effect on anyone unless they are using vim, in which case it is actually helpful.

Reply:
> Can this be in config file? as `jerryscript-core.h` also need this.

Unfortunately no, it's effective only for the file that contains the directive.

Reply:
> Bunch of git-ignores for a lot of editors

I'm not sure what you mean by that. Calling gitignore an IDE/editor-specific file would be an overstatement. If you mean that gitignore contains lines that mention IDE-specific files, well, then yes. But that means two things: those files are not included in the code base & those details are kept at a single place (i.e., in the gitignore file).

> generated compile_commands.json to enable code completion in vscode, etc.

Again, these are kept in separate files, not mixed into the sources.

As for vim, it should be possible for that editor too to keep its settings in a single file. I'm not a user of vim but did some research and found several options:
- https://vim.fandom.com/wiki/Project_specific_settings
- https://github.com/LucHermitte/local_vimrc
- https://github.com/embear/vim-localvimrc

And regarding "what the fuss is about", I completely agree with https://stackoverflow.com/questions/456792/vim-apply-settings-on-files-in-directory/456889#456889

>DRY: with modelines, a setting needs to be repeated in every file, if there are too many things to set or tunings to change, it will quickly become difficult to maintain, moreover, it will require the use of a template-expander plugin (which you should consider if you have several vimmers in your project).
Not every one uses vim to develop. I don't want to be bothered by other people editor settings, why should I parasite theirs?
It's easier to ask vimmers to install a same plugin, instead of asking them to copy-paste, and maintain, the same lines in their .vimrc
The settings can be saved with the other project files (cvs/svn/git/whatever)

Reply:
I'm not calling gitignore IDE specific, I'm pointing out that outgitignore is already bloated with edit specific ignore lines. These wouldn't be necessary, as the don't pertain to the code base, and people could create ignore them locally in their repository, depending on what they do ro do not use. Nevertheless, we keep it in the repository to help developers and make their life easier.

However back to the point.
This modeline is not needed for every file in the source tree. What this modeline does, just to make things clear, is it defines how folds can be created for doxygen comment groups, so that they can optionally be collapsed, to make reading the file easier. It isn't a typical 'editor setting', that configures the editor in just the right way, like setting tab width etc. These directives are tied to doxygen comment groups, and don't need any maintenance as these don't change.

The reason I added them specifically to these headers is that these files contain a lot of text, and will contain more in the near future. Allowing to collapse groups makes reading the file a lot easier.

Think of it this way. If I were to add a hard-to-understand line of code, I would add a line of comment describing it.
This comment would contain useful information which the developer can use to understand the code more easily. This comment does exactly the same, it contains helpful information for reading the file, with the one exception, that it is not in some textual human language. Even if someone uses vim, they are not obligated to use it, they can decide whether they want to parse modelines or not.

For the majority of users this comment will affect nothing. It requires no maintenance, it is not needed to be added to every source file in the project, in fact it is not needed at all. However for some small percentage of people it will be helpful to have it in these files specifically, as they can make it easier to go through the wall of text.

Reply:
I did look up the meaning of the modeline, so it is clear what it is achieving. And I also agree that it may be worth adding such developer-assisting settings or configurations to the code base. Where I do not agree is where such settings should be placed. Single tool-specific (e.g., .clang-format) or non-code (e.g., .gitignore) files: yes. Public headers: less so. (BTW, actually even internal files use this doxygen syntax for grouping parts of large files (e.g., js-parser-internal.h or jmem-allocator-internal.h). So this _is_ a project-wide setting, not only related to two public headers.)

Reply:
You are correct that that internal sources also contain grouping directives, however that is also that would cause problems if we would set this project wide.

Source files usually only contain a single group for the entire file, and collapsing the whole source text isn't very helpful. Also in source files you'd normally prefer a different mode of folding, for example by functions or by blocks. So this setting only really makes sense for these headers, where a lot of groups are created.

One other difference is that tool specific config files are only really used by developers. This setting is not restricted to development purposes. For example, if one were to package the header, then someone reading it could make use of the modeline, whereas if it was set in a config file, that would most likely not be included.